File: ios-sdk/Sources/GazePointSDK/GazeTracker.swift
import Foundation
import Vision
import ARKit
import CoreML
import UIKit
import Accelerate

/// Enhanced Gaze Tracker for iOS with Vision and ARKit support
@available(iOS 16.0, *)
public class GazeTracker {
    
    // MARK: - Types
    
    public struct GazeResult {
        public let gazePoint: CGPoint
        public let confidence: Float
        public let isBlinking: Bool
        public let headPose: HeadPose
        public let timestamp: TimeInterval
        
        public init(gazePoint: CGPoint, confidence: Float, isBlinking: Bool, headPose: HeadPose, timestamp: TimeInterval) {
            self.gazePoint = gazePoint
            self.confidence = confidence
            self.isBlinking = isBlinking
            self.headPose = headPose
            self.timestamp = timestamp
        }
    }
    
    public struct HeadPose {
        public let pitch: Float // Nodding up/down
        public let yaw: Float   // Turning left/right
        public let roll: Float  // Tilting left/right
        
        public init(pitch: Float, yaw: Float, roll: Float) {
            self.pitch = pitch
            self.yaw = yaw
            self.roll = roll
        }
    }
    
    public struct CalibrationData: Codable {
        public var offsetX: Float
        public var offsetY: Float
        public var scaleX: Float
        public var scaleY: Float
        public var rotationCompensation: Float
        
        public init(offsetX: Float = 0, offsetY: Float = 0, scaleX: Float = 1, scaleY: Float = 1, rotationCompensation: Float = 0) {
            self.offsetX = offsetX
            self.offsetY = offsetY
            self.scaleX = scaleX
            self.scaleY = scaleY
            self.rotationCompensation = rotationCompensation
        }
    }
    
    // MARK: - Properties
    
    private let smoothingFactor: Float = 0.3
    private let minConfidenceThreshold: Float = 0.5
    private let blinkThreshold: Float = 0.3
    private let velocityThreshold: Float = 100.0
    
    private var lastGazePoint: CGPoint?
    private var calibrationData: CalibrationData?
    private var isCalibrated: Bool = false
    private var kalmanFilter: KalmanFilter
    private let performanceMonitor: PerformanceMonitor
    
    private lazy var faceDetectionRequest: VNDetectFaceLandmarksRequest = {
        let request = VNDetectFaceLandmarksRequest()
        request.revision = VNDetectFaceLandmarksRequestRevision3
        return request
    }()
    
    // MARK: - Initialization
    
    public init() {
        self.kalmanFilter = KalmanFilter()
        self.performanceMonitor = PerformanceMonitor()
    }
    
    // MARK: - Public Methods
    
    /// Calculate gaze point from a CVPixelBuffer
    public func calculateGazePoint(from pixelBuffer: CVPixelBuffer, orientation: CGImagePropertyOrientation = .up) -> GazeResult? {
        let startTime = performanceMonitor.startFrame()
        defer {
            performanceMonitor.endFrame(startTime: startTime)
        }
        
        let handler = VNImageRequestHandler(cvPixelBuffer: pixelBuffer, orientation: orientation, options: [:])
        
        do {
            try handler.perform([faceDetectionRequest])
            
            guard let observations = faceDetectionRequest.results,
                  let faceObservation = observations.first else {
                return nil
            }
            
            let result = processGaze(from: faceObservation)
            return result
            
        } catch {
            print("Error performing face detection: \(error)")
            return nil
        }
    }
    
    /// Calculate gaze point from a UIImage
    public func calculateGazePoint(from image: UIImage) -> GazeResult? {
        guard let cgImage = image.cgImage else { return nil }
        
        let startTime = performanceMonitor.startFrame()
        defer {
            performanceMonitor.endFrame(startTime: startTime)
        }
        
        let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        
        do {
            try handler.perform([faceDetectionRequest])
            
            guard let observations = faceDetectionRequest.results,
                  let faceObservation = observations.first else {
                return nil
            }
            
            return processGaze(from: faceObservation)
            
        } catch {
            print("Error performing face detection: \(error)")
            return nil
        }
    }
    
    /// Calibrate the gaze tracker with known screen points
    public func calibrate(calibrationPoints: [(expected: CGPoint, actual: CGPoint)]) {
        guard calibrationPoints.count >= 3 else {
            print("Need at least 3 calibration points")
            return
        }
        
        var sumOffsetX: Float = 0
        var sumOffsetY: Float = 0
        var sumScaleX: Float = 0
        var sumScaleY: Float = 0
        
        for (expected, actual) in calibrationPoints {
            sumOffsetX += Float(expected.x - actual.x)
            sumOffsetY += Float(expected.y - actual.y)
            
            if actual.x != 0 {
                sumScaleX += Float(expected.x / actual.x)
            }
            if actual.y != 0 {
                sumScaleY += Float(expected.y / actual.y)
            }
        }
        
        let count = Float(calibrationPoints.count)
        calibrationData = CalibrationData(
            offsetX: sumOffsetX / count,
            offsetY: sumOffsetY / count,
            scaleX: sumScaleX / count,
            scaleY: sumScaleY / count
        )
        
        isCalibrated = true
        print("Calibration completed: \(String(describing: calibrationData))")
    }
    
    /// Reset calibration
    public func resetCalibration() {
        calibrationData = nil
        isCalibrated = false
        lastGazePoint = nil
        kalmanFilter.reset()
    }
    
    /// Get current performance metrics
    public func getPerformanceMetrics() -> PerformanceMonitor.PerformanceMetrics {
        return performanceMonitor.getMetrics()
    }
    
    // MARK: - Private Methods
    
    private func processGaze(from faceObservation: VNFaceObservation) -> GazeResult? {
        guard let landmarks = faceObservation.landmarks else {
            return nil
        }
        
        // Get eye landmarks
        guard let leftEye = landmarks.leftEye,
              let rightEye = landmarks.rightEye else {
            return nil
        }
        
        // Detect blink
        let isBlinking = detectBlink(leftEye: leftEye, rightEye: rightEye, faceObservation: faceObservation)
        
        // Calculate head pose
        let headPose = calculateHeadPose(from: faceObservation)
        
        // Calculate gaze vector
        let gazeVector = calculateGazeVector(leftEye: leftEye, rightEye: rightEye, headPose: headPose)
        
        // Apply calibration if available
        let calibratedVector = applyCalibration(to: gazeVector)
        
        // Map to screen coordinates
        let screenSize = UIScreen.main.bounds.size
        let screenPoint = mapGazeVectorToScreen(gazeVector: calibratedVector, headPose: headPose, screenSize: screenSize)
        
        // Apply Kalman filter
        let filteredPoint = kalmanFilter.update(measurement: screenPoint)
        
        // Apply adaptive smoothing
        let smoothedPoint = applyAdaptiveSmoothing(currentPoint: filteredPoint)
        
        // Calculate confidence
        let confidence = calculateConfidence(faceObservation: faceObservation, isBlinking: isBlinking)
        
        lastGazePoint = smoothedPoint
        
        return GazeResult(
            gazePoint: smoothedPoint,
            confidence: confidence,
            isBlinking: isBlinking,
            headPose: headPose,
            timestamp: Date().timeIntervalSince1970
        )
    }
    
    private func calculateGazeVector(leftEye: VNFaceLandmarkRegion2D, rightEye: VNFaceLandmarkRegion2D, headPose: HeadPose) -> CGPoint {
        let leftPoints = leftEye.normalizedPoints
        let rightPoints = rightEye.normalizedPoints
        
        guard !leftPoints.isEmpty, !rightPoints.isEmpty else {
            return .zero
        }
        
        // Calculate eye centers
        let leftCenter = averagePoint(points: leftPoints)
        let rightCenter = averagePoint(points: rightPoints)
        
        // Calculate eye midpoint
        let eyeMidX = (leftCenter.x + rightCenter.x) / 2
        let eyeMidY = (leftCenter.y + rightCenter.y) / 2
        
        // Calculate base gaze vector
        var gazeX = rightCenter.x - leftCenter.x
        var gazeY = rightCenter.y - leftCenter.y
        
        // Compensate for head rotation
        gazeX += CGFloat(headPose.yaw) * 0.005
        gazeY += CGFloat(headPose.pitch) * 0.005
        
        // Normalize
        let magnitude = sqrt(gazeX * gazeX + gazeY * gazeY)
        if magnitude > 0 {
            gazeX /= magnitude
            gazeY /= magnitude
        }
        
        return CGPoint(x: gazeX, y: gazeY)
    }
    
    private func calculateHeadPose(from faceObservation: VNFaceObservation) -> HeadPose {
        let pitch = faceObservation.pitch?.floatValue ?? 0
        let yaw = faceObservation.yaw?.floatValue ?? 0
        let roll = faceObservation.roll?.floatValue ?? 0
        
        return HeadPose(
            pitch: pitch,
            yaw: yaw,
            roll: roll
        )
    }
    
    private func detectBlink(leftEye: VNFaceLandmarkRegion2D, rightEye: VNFaceLandmarkRegion2D, faceObservation: VNFaceObservation) -> Bool {
        // Calculate eye aspect ratios
        let leftEAR = calculateEyeAspectRatio(eye: leftEye)
        let rightEAR = calculateEyeAspectRatio(eye: rightEye)
        
        let avgEAR = (leftEAR + rightEAR) / 2
        
        return avgEAR < blinkThreshold
    }
    
    private func calculateEyeAspectRatio(eye: VNFaceLandmarkRegion2D) -> Float {
        let points = eye.normalizedPoints
        guard points.count >= 6 else { return 1.0 }
        
        // Simplified EAR calculation
        // EAR = (vertical distance) / (horizontal distance)
        let vertical1 = distance(from: points[1], to: points[5])
        let vertical2 = distance(from: points[2], to: points[4])
        let horizontal = distance(from: points[0], to: points[3])
        
        if horizontal == 0 { return 1.0 }
        
        let ear = Float((vertical1 + vertical2) / (2.0 * horizontal))
        return ear
    }
    
    private func mapGazeVectorToScreen(gazeVector: CGPoint, headPose: HeadPose, screenSize: CGSize) -> CGPoint {
        // Apply head pose compensation
        let yawFactor: CGFloat = 1.0 + (CGFloat(abs(headPose.yaw)) / 30.0) * 0.2
        let pitchFactor: CGFloat = 1.0 + (CGFloat(abs(headPose.pitch)) / 30.0) * 0.2
        
        // Map to screen coordinates
        var screenX = (screenSize.width / 2) + (gazeVector.x * (screenSize.width / 2) * yawFactor)
        var screenY = (screenSize.height / 2) - (gazeVector.y * (screenSize.height / 2) * pitchFactor)
        
        // Clamp to screen bounds
        screenX = max(0, min(screenSize.width, screenX))
        screenY = max(0, min(screenSize.height, screenY))
        
        return CGPoint(x: screenX, y: screenY)
    }
    
    private func applyCalibration(to gazeVector: CGPoint) -> CGPoint {
        guard isCalibrated, let calibration = calibrationData else {
            return gazeVector
        }
        
        return CGPoint(
            x: CGFloat(gazeVector.x * CGFloat(calibration.scaleX) + CGFloat(calibration.offsetX)),
            y: CGFloat(gazeVector.y * CGFloat(calibration.scaleY) + CGFloat(calibration.offsetY))
        )
    }
    
    private func applyAdaptiveSmoothing(currentPoint: CGPoint) -> CGPoint {
        guard let lastPoint = lastGazePoint else {
            return currentPoint
        }
        
        // Calculate velocity
        let dx = currentPoint.x - lastPoint.x
        let dy = currentPoint.y - lastPoint.y
        let velocity = sqrt(dx * dx + dy * dy)
        
        // Adaptive smoothing factor
        let adaptiveFactor: CGFloat
        if velocity > CGFloat(velocityThreshold) {
            adaptiveFactor = CGFloat(smoothingFactor) * 0.5
        } else {
            adaptiveFactor = CGFloat(smoothingFactor)
        }
        
        return CGPoint(
            x: lastPoint.x + (currentPoint.x - lastPoint.x) * adaptiveFactor,
            y: lastPoint.y + (currentPoint.y - lastPoint.y) * adaptiveFactor
        )
    }
    
    private func calculateConfidence(faceObservation: VNFaceObservation, isBlinking: Bool) -> Float {
        var confidence = faceObservation.confidence
        
        // Reduce confidence if blinking
        if isBlinking {
            confidence *= 0.3
        }
        
        // Check face quality
        if let boundingBox = Optional(faceObservation.boundingBox) {
            let faceArea = boundingBox.width * boundingBox.height
            if faceArea < 0.05 { // Face too small
                confidence *= 0.7
            }
        }
        
        return max(0, min(1, confidence))
    }
    
    // MARK: - Helper Methods
    
    private func averagePoint(points: [CGPoint]) -> CGPoint {
        guard !points.isEmpty else { return .zero }
        
        var sumX: CGFloat = 0
        var sumY: CGFloat = 0
        
        for point in points {
            sumX += point.x
            sumY += point.y
        }
        
        return CGPoint(x: sumX / CGFloat(points.count), y: sumY / CGFloat(points.count))
    }
    
    private func distance(from p1: CGPoint, to p2: CGPoint) -> CGFloat {
        let dx = p2.x - p1.x
        let dy = p2.y - p1.y
        return sqrt(dx * dx + dy * dy)
    }
}

File: app/src/main/java/com/facedetection/gaze/GazeTracker.kt
package com.facedetection.gaze

import android.content.Context
import android.graphics.PointF
import android.util.DisplayMetrics
import android.util.Log
import com.google.ar.sceneform.math.Vector3
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceLandmark
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.sqrt

/**
 * Enhanced Gaze Tracker with improved accuracy and calibration support.
 * 
 * Features:
 * - Kalman filtering for smooth tracking
 * - Multi-point calibration support
 * - Head pose compensation
 * - Adaptive smoothing based on movement speed
 * - Eye aspect ratio for blink detection
 */
class GazeTracker(private val context: Context) {
    
    companion object {
        private const val TAG = "GazeTracker"
        private const val SMOOTHING_FACTOR = 0.3f
        private const val MIN_CONFIDENCE_THRESHOLD = 0.5f
        private const val BLINK_EAR_THRESHOLD = 0.2f
        private const val VELOCITY_THRESHOLD = 100f
    }
    
    private var lastGazePoint: PointF? = null
    private var calibrationData: CalibrationData? = null
    private var isCalibrated = false
    private var blinkDetected = false
    private val kalmanFilter = KalmanFilter()
    
    data class CalibrationData(
        val offsetX: Float = 0f,
        val offsetY: Float = 0f,
        val scaleX: Float = 1f,
        val scaleY: Float = 1f,
        val rotationCompensation: Float = 0f
    )
    
    data class GazeResult(
        val gazePoint: PointF,
        val confidence: Float,
        val isBlinking: Boolean,
        val headPose: HeadPose
    )
    
    data class HeadPose(
        val pitch: Float,
        val yaw: Float,
        val roll: Float
    )
    
    /**
     * Calculate gaze point from detected face with improved accuracy.
     */
    fun calculateGazePoint(face: Face): GazeResult? {
        val leftEyeLandmark = face.getLandmark(FaceLandmark.LEFT_EYE)
        val rightEyeLandmark = face.getLandmark(FaceLandmark.RIGHT_EYE)
        
        if (leftEyeLandmark == null || rightEyeLandmark == null) {
            Log.w(TAG, "Eye landmarks not detected")
            return null
        }
        
        val leftEyePosition = leftEyeLandmark.position
        val rightEyePosition = rightEyeLandmark.position
        
        // Detect blink using Eye Aspect Ratio
        blinkDetected = detectBlink(face)
        
        // Calculate head pose
        val headPose = calculateHeadPose(face)
        
        // Calculate raw gaze vector
        val gazeVector = calculateGazeVector(leftEyePosition, rightEyePosition, headPose)
        
        // Apply calibration if available
        val calibratedVector = if (isCalibrated && calibrationData != null) {
            applyCalibration(gazeVector, calibrationData!!)
        } else {
            gazeVector
        }
        
        // Map to screen coordinates
        val screenPoint = mapGazeVectorToScreenCoordinates(calibratedVector, headPose)
        
        // Apply Kalman filter for smoothing
        val filteredPoint = kalmanFilter.update(screenPoint)
        
        // Apply adaptive smoothing based on velocity
        val smoothedPoint = applyAdaptiveSmoothing(filteredPoint)
        
        // Calculate confidence based on multiple factors
        val confidence = calculateConfidence(face, leftEyePosition, rightEyePosition)
        
        lastGazePoint = smoothedPoint
        
        return GazeResult(
            gazePoint = smoothedPoint,
            confidence = confidence,
            isBlinking = blinkDetected,
            headPose = headPose
        )
    }
    
    /**
     * Calculate gaze vector with head pose compensation.
     */
    private fun calculateGazeVector(
        leftEyePosition: PointF,
        rightEyePosition: PointF,
        headPose: HeadPose
    ): Vector3 {
        // Calculate eye midpoint
        val eyeMidPoint = PointF(
            (leftEyePosition.x + rightEyePosition.x) / 2f,
            (leftEyePosition.y + rightEyePosition.y) / 2f
        )
        
        // Calculate base gaze vector
        val baseVector = Vector3(
            rightEyePosition.x - leftEyePosition.x,
            rightEyePosition.y - leftEyePosition.y,
            0f
        )
        
        // Compensate for head rotation
        val compensatedX = baseVector.x + headPose.yaw * 0.5f
        val compensatedY = baseVector.y + headPose.pitch * 0.5f
        
        val gazeVector = Vector3(compensatedX, compensatedY, 0f)
        
        // Normalize
        val magnitude = sqrt(gazeVector.x * gazeVector.x + gazeVector.y * gazeVector.y)
        if (magnitude > 0) {
            return Vector3(gazeVector.x / magnitude, gazeVector.y / magnitude, 0f)
        }
        
        return gazeVector
    }
    
    /**
     * Calculate head pose from face landmarks.
     */
    private fun calculateHeadPose(face: Face): HeadPose {
        val pitch = face.headEulerAngleX // Nodding up and down
        val yaw = face.headEulerAngleY   // Turning left and right
        val roll = face.headEulerAngleZ  // Tilting left and right
        
        return HeadPose(pitch, yaw, roll)
    }
    
    /**
     * Detect blink using Eye Aspect Ratio (EAR).
     */
    private fun detectBlink(face: Face): Boolean {
        // Get eye landmarks for more accurate blink detection
        val leftEye = face.getLandmark(FaceLandmark.LEFT_EYE)
        val rightEye = face.getLandmark(FaceLandmark.RIGHT_EYE)
        
        if (leftEye == null || rightEye == null) return false
        
        // Use eye open probability if available
        val leftEyeOpen = face.leftEyeOpenProbability ?: 1f
        val rightEyeOpen = face.rightEyeOpenProbability ?: 1f
        
        val avgEyeOpen = (leftEyeOpen + rightEyeOpen) / 2f
        
        return avgEyeOpen < BLINK_EAR_THRESHOLD
    }
    
    /**
     * Map gaze vector to screen coordinates with head pose compensation.
     */
    private fun mapGazeVectorToScreenCoordinates(
        gazeVector: Vector3,
        headPose: HeadPose
    ): PointF {
        val displayMetrics = context.resources.displayMetrics
        val screenWidth = displayMetrics.widthPixels.toFloat()
        val screenHeight = displayMetrics.heightPixels.toFloat()
        
        // Apply head pose compensation
        val yawFactor = 1f + (abs(headPose.yaw) / 30f) * 0.2f
        val pitchFactor = 1f + (abs(headPose.pitch) / 30f) * 0.2f
        
        // Map to screen coordinates with compensation
        val screenX = (screenWidth / 2f) + (gazeVector.x * (screenWidth / 2f) * yawFactor)
        val screenY = (screenHeight / 2f) - (gazeVector.y * (screenHeight / 2f) * pitchFactor)
        
        // Clamp to screen bounds
        val clampedX = screenX.coerceIn(0f, screenWidth)
        val clampedY = screenY.coerceIn(0f, screenHeight)
        
        return PointF(clampedX, clampedY)
    }
    
    /**
     * Apply calibration data to gaze vector.
     */
    private fun applyCalibration(gazeVector: Vector3, calibration: CalibrationData): Vector3 {
        return Vector3(
            gazeVector.x * calibration.scaleX + calibration.offsetX,
            gazeVector.y * calibration.scaleY + calibration.offsetY,
            gazeVector.z
        )
    }
    
    /**
     * Apply adaptive smoothing based on gaze velocity.
     */
    private fun applyAdaptiveSmoothing(currentPoint: PointF): PointF {
        val lastPoint = lastGazePoint ?: return currentPoint
        
        // Calculate velocity
        val dx = currentPoint.x - lastPoint.x
        val dy = currentPoint.y - lastPoint.y
        val velocity = sqrt(dx * dx + dy * dy)
        
        // Adaptive smoothing factor based on velocity
        val adaptiveFactor = if (velocity > VELOCITY_THRESHOLD) {
            SMOOTHING_FACTOR * 0.5f // Less smoothing for fast movements
        } else {
            SMOOTHING_FACTOR // More smoothing for slow movements
        }
        
        return PointF(
            lastPoint.x + (currentPoint.x - lastPoint.x) * adaptiveFactor,
            lastPoint.y + (currentPoint.y - lastPoint.y) * adaptiveFactor
        )
    }
    
    /**
     * Calculate confidence score based on multiple factors.
     */
    private fun calculateConfidence(
        face: Face,
        leftEyePosition: PointF,
        rightEyePosition: PointF
    ): Float {
        var confidence = 1.0f
        
        // Face detection confidence
        confidence *= (face.trackingId ?: 0).let { if (it > 0) 1.0f else 0.8f }
        
        // Eye position stability
        lastGazePoint?.let { last ->
            val eyeMidX = (leftEyePosition.x + rightEyePosition.x) / 2f
            val eyeMidY = (leftEyePosition.y + rightEyePosition.y) / 2f
            val stability = 1f - (abs(last.x - eyeMidX) + abs(last.y - eyeMidY)) / 1000f
            confidence *= stability.coerceIn(0.5f, 1.0f)
        }
        
        // Blink penalty
        if (blinkDetected) {
            confidence *= 0.3f
        }
        
        return confidence.coerceIn(0f, 1f)
    }
    
    /**
     * Calibrate the gaze tracker with known screen points.
     */
    fun calibrate(calibrationPoints: List<Pair<PointF, PointF>>) {
        if (calibrationPoints.size < 3) {
            Log.w(TAG, "Need at least 3 calibration points")
            return
        }
        
        // Calculate calibration offsets and scales
        var sumOffsetX = 0f
        var sumOffsetY = 0f
        var sumScaleX = 0f
        var sumScaleY = 0f
        
        calibrationPoints.forEach { (expected, actual) ->
            sumOffsetX += expected.x - actual.x
            sumOffsetY += expected.y - actual.y
            
            if (actual.x != 0f) sumScaleX += expected.x / actual.x
            if (actual.y != 0f) sumScaleY += expected.y / actual.y
        }
        
        val count = calibrationPoints.size
        calibrationData = CalibrationData(
            offsetX = sumOffsetX / count,
            offsetY = sumOffsetY / count,
            scaleX = sumScaleX / count,
            scaleY = sumScaleY / count
        )
        
        isCalibrated = true
        Log.i(TAG, "Calibration completed: $calibrationData")
    }
    
    /**
     * Reset calibration.
     */
    fun resetCalibration() {
        calibrationData = null
        isCalibrated = false
        lastGazePoint = null
        kalmanFilter.reset()
    }
    
    /**
     * Simple Kalman Filter for smoothing gaze points.
     */
    private class KalmanFilter {
        private var estimateX = 0f
        private var estimateY = 0f
        private var errorCovarianceX = 1f
        private var errorCovarianceY = 1f
        
        private val processNoise = 0.01f
        private val measurementNoise = 0.1f
        
        fun update(measurement: PointF): PointF {
            // Prediction (assume no change)
            val predictedErrorCovX = errorCovarianceX + processNoise
            val predictedErrorCovY = errorCovarianceY + processNoise
            
            // Update
            val kalmanGainX = predictedErrorCovX / (predictedErrorCovX + measurementNoise)
            val kalmanGainY = predictedErrorCovY / (predictedErrorCovY + measurementNoise)
            
            estimateX += kalmanGainX * (measurement.x - estimateX)
            estimateY += kalmanGainY * (measurement.y - estimateY)
            
            errorCovarianceX = (1 - kalmanGainX) * predictedErrorCovX
            errorCovarianceY = (1 - kalmanGainY) * predictedErrorCovY
            
            return PointF(estimateX, estimateY)
        }
        
        fun reset() {
            estimateX = 0f
            estimateY = 0f
            errorCovarianceX = 1f
            errorCovarianceY = 1f
        }
    }
}
